prover/lang/kore-lang.md

```diff
@@ -425,36 +453,66 @@ and values, passed to K's substitute.
-  rule subst(\equals(ARG1, ARG2):Pattern, X, V)
-    => \equals(subst(ARG1, X, V), subst(ARG2, X, V)):Pattern
-  rule subst(\not(ARG):Pattern, X, V) => \not(subst(ARG, X, V)):Pattern
-  rule subst(\and(ARG):Pattern, X, V) => \and(ARG[X/V]):Pattern
-  rule subst(\or(ARG):Pattern, X, V) => \or(ARG[X/V]):Pattern
-  rule subst(\implies(LHS, RHS):Pattern, X, V)
-    => \implies(subst(LHS, X, V), subst(RHS, X, V)):Pattern
-  rule subst(\forall { E } C, X, V) => \forall { E } subst(C, X, V)
-  rule subst(\exists { E } C, X, V) => \exists { E } subst(C, X, V)
-
-  rule subst(S:Symbol, X, V) => S
+  syntax PatternOrVarName ::= Pattern | VariableName
+
+  // capture-avoiding substitution
+  syntax Pattern ::= casubst(Pattern, PatternOrVarName, Pattern)    [function]
+
+  rule casubst(P, V, S) => #subst(true, P, V, S)
+
+  // non-capture-avoiding substitution
+  syntax Pattern ::= subst(Pattern, PatternOrVarName, Pattern)    [function, klabel(subst)]
+
+  rule subst(P, V, S) => #subst(false, P, V, S)
+
+  syntax Pattern ::= #subst(Bool, Pattern, PatternOrVarName, Pattern)    [function]
+
+  rule #subst(_, T,T,V) => V // We allow substitution over arbitary patterns
+  rule #subst(_, X{_}, X:VariableName, V) => V
+  rule #subst(_, X{S}, Y:VariableName, V) => X{S} requires X =/=K Y
+  rule #subst(_, X:Variable,Y:Variable,V) => X requires X =/=K Y
+  rule #subst(_, I:Int, X, V) => I
+  rule #subst(_, \top(),_,_)=> \top()
+  rule #subst(_, \bottom(),_,_) => \bottom()
+  rule #subst(CA, \equals(ARG1, ARG2):Pattern, X, V)
+    => \equals(#subst(CA, ARG1, X, V), #subst(CA, ARG2, X, V)):Pattern
+  rule #subst(CA, \not(ARG):Pattern, X, V) => \not(#subst(CA, ARG, X, V)):Pattern
+  rule #subst(CA, \and(ARG):Pattern, X, V) => \and(ARG[CA, X/V]):Pattern
+  rule #subst(CA, \or(ARG):Pattern, X, V) => \or(ARG[CA, X/V]):Pattern
+  rule #subst(CA, \implies(LHS, RHS):Pattern, X, V)
+    => \implies(#subst(CA, LHS, X, V), #subst(CA, RHS, X, V)):Pattern
+  rule #subst(CA:Bool, \forall { E } C, X, V)
+    => #if CA andBool (
+         X in E orBool X in PatternsToVariableNameSet(E))
+       #then \forall { E } C
+       #else \forall { E } #subst(CA, C, X, V)
+       #fi
+  rule #subst(CA, \exists { E } C, X, V)
+    => #if CA andBool (
+         X in E orBool X in PatternsToVariableNameSet(E))
+       #then \exists { E } C
+       #else \exists { E } #subst(CA, C, X, V)
+       #fi
+
+  rule #subst(_, S:Symbol, X, V) => S
     requires S =/=K X
-  rule subst(S:Symbol(ARGS:Patterns) #as T:Pattern, X, V) => S(ARGS[X/V])
+  rule #subst(CA, S:Symbol(ARGS:Patterns) #as T:Pattern, X, V) => S(ARGS[CA, X/V])
     requires T =/=K X
 
-  rule subst(implicationContext(CTX, RHS), X, V)
-    => implicationContext(subst(CTX,X,V), subst(RHS,X,V)):Pattern
+  rule #subst(CA, implicationContext(CTX, RHS), X, V)
+    => implicationContext(#subst(CA,CTX,X,V), #subst(CA,RHS,X,V)):Pattern
 
-  rule subst(R,_,_) => R [owise]
+  rule #subst(_, R,_,_) => R [owise]
```

Comment:
Ugh. This has become pretty complex. We really need unit tests. Wish we had a good option for that.